src/rust/engine/fs/store/src/remote.rs

```diff
@@ -87,31 +120,72 @@ impl ByteStore {
   async fn with_byte_stream_client<
     Value: Send,
     Fut: std::future::Future<Output = Result<Value, String>>,
-    F: Fn(bazel_protos::bytestream_grpc::ByteStreamClient) -> Fut,
+    F: Fn(ByteStreamClient<tonic::transport::Channel>) -> Fut,
   >(
     &self,
     f: F,
   ) -> Result<Value, String> {
-    retry::all_errors_immediately(&self.serverset, self.rpc_attempts, move |channel| {
-      f(bazel_protos::bytestream_grpc::ByteStreamClient::new(
-        channel,
-      ))
-    })
-    .await
+    // retry::all_errors_immediately(&self.serverset, self.rpc_attempts, move |channel| {
+    //   f(bazel_protos::bytestream_grpc::ByteStreamClient::new(
+    //     channel,
+    //   ))
+    // })
+    // .await
+
+    if self.headers.is_empty() {
+      let client = ByteStreamClient::new(self.channel.clone());
+      f(client).await
+    } else {
+      let headers = self.headers.clone();
+      let client =
+        ByteStreamClient::with_interceptor(self.channel.clone(), move |mut req: Request<()>| {
+          let metadata = req.metadata_mut();
+          for (key, value) in &headers {
+            let key_ascii = AsciiMetadataKey::from_str(key.as_str()).unwrap();
+            let value_ascii = AsciiMetadataValue::from_str(value.as_str()).unwrap();
+            metadata.insert(key_ascii, value_ascii);
+          }
+          Ok(req)
+        });
+      f(client).await
+    }
   }
 
   async fn with_cas_client<
     Value: Send,
     Fut: std::future::Future<Output = Result<Value, String>>,
-    F: Fn(bazel_protos::remote_execution_grpc::ContentAddressableStorageClient) -> Fut,
+    F: Fn(ContentAddressableStorageClient<Channel>) -> Fut,
   >(
     &self,
     f: F,
   ) -> Result<Value, String> {
-    retry::all_errors_immediately(&self.serverset, self.rpc_attempts, move |channel| {
-      f(bazel_protos::remote_execution_grpc::ContentAddressableStorageClient::new(channel))
-    })
-    .await
+    // retry::all_errors_immediately(&self.serverset, self.rpc_attempts, move |channel| {
```

Comment:
Commented out since needs research on how it should interact with Tonic.

Reply:
I switched to using a common interceptor that is cloned for each of these. I'll switch to making the clients once in a separate PR.